plugins/browser-plugin-snowplow-ecommerce/src/ua/api.ts

```diff
@@ -0,0 +1,98 @@
+import {
+  trackProductView,
+  trackProductListView,
+  trackProductListClick,
+  trackCheckoutStep,
+  trackTransaction,
+  trackAddToCart,
+  trackRemoveFromCart,
+} from '../api';
+import { CheckoutStep } from '../types';
+import {
+  AddToCartEvent,
+  CheckoutStepEvent,
+  ProductDetailEvent,
+  ProductListClickEvent,
+  ProductListViewEvent,
+  PromoClickEvent,
+  PromoViewEvent,
+  PurchaseEvent,
+  RemoveFromCartEvent,
+  UAEnhancedEcommerceObject,
+} from './types';
+
+import { transformUAProductsToSPProducts } from './utils';
+
+const DEFAULT_PAYMENT_METHOD = 'unknown';
+
+export function trackEEPromoView(ecommerce: UAEnhancedEcommerceObject & PromoViewEvent) {}
+export function trackEEPromoClick(ecommerce: UAEnhancedEcommerceObject & PromoClickEvent) {}
+
+export function trackEEAddToCart(ecommerce: UAEnhancedEcommerceObject & AddToCartEvent, finalCartValue: number) {
+  const products = transformUAProductsToSPProducts({
+    currencyCode: ecommerce.currencyCode,
+    products: ecommerce.add.products,
+  });
+  trackAddToCart({ products, total_value: finalCartValue, currency: ecommerce.currencyCode });
+}
+
+export function trackEERemoveFromCart(
+  ecommerce: UAEnhancedEcommerceObject & RemoveFromCartEvent,
+  finalCartValue: number
+) {
+  const products = transformUAProductsToSPProducts({
+    currencyCode: ecommerce.currencyCode,
+    products: ecommerce.remove.products,
+  });
+  trackRemoveFromCart({ products, total_value: finalCartValue, currency: ecommerce.currencyCode });
+}
+
+export function trackEEProductDetail(ecommerce: UAEnhancedEcommerceObject & ProductDetailEvent) {
+  const products = transformUAProductsToSPProducts({
+    currencyCode: ecommerce.currencyCode,
+    products: ecommerce.detail.products,
+  });
+  trackProductView({ ...products[0] });
+}
+
+export function trackEEProductListView(ecommerce: UAEnhancedEcommerceObject & ProductListViewEvent) {
+  const products = transformUAProductsToSPProducts({
+    currencyCode: ecommerce.currencyCode,
+    products: ecommerce.impressions,
+  });
+  trackProductListView({ name: ecommerce.impressions[0].list as string, products });
```

Comment:
Also I'd generally prefer `val!` to `val as string` as I think the intent is clearer: (despite the type we really expect this to be defined at this point) vs (shut up typescript, this is a string no matter what I said before)

Reply:
In my view if the list is empty on the `trackEnhancedEcommerceProductListView` call, then I would think it should fail as a bad event on the pipeline. This can then unearth any product list issues that were previously uknown to the user due to GA not indicating an otherwise invalid event.

On the note of `val!` I agree as well. Fixed